pkg/lib/cockpit/location.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { url_root, calculate_application } from './_internal/location-utils';
+
+/* HACK: Mozilla will unescape 'window.location.hash' before returning
+* it, which is broken.
+*
+* https://bugzilla.mozilla.org/show_bug.cgi?id=135309
+*/
+export function get_window_location_hash() {
+    return (window.location.href.split('#')[1] || '');
+}
+
+type Options = { [name: string]: string | Array<string> };
```

Comment:
Please help me to understand this: From the API point of view, the options can be pretty much anything, right? Are strings and string lists the only possible options, or are there e.g. numerical or boolean (presence/absence) options? The lone `option2` in https://cockpit-project.org/guide/latest/cockpit-location.html reads like a boolean-ish thing.

Are these the url-decoded attributes from `Location`'s `href`?

Reply:
So I did a deep dive here, this cannot be numerical / boolean.

```
https://localhost:9090/system#/foo?sort=&repo=Core-Staging&repo=Core-Testing&repo=Extra&q=cinnamon%2Cwejfi&maintainer=torxed&flagged=Flagged
```

window.location.options becomes
```
{
  "sort": "",
  "repo": [
    "Core-Staging",
    "Core-Testing",
    "Extra"
  ],
  "q": "cinnamon,wejfi",
  "maintainer": "torxed",
  "flagged": "Flagged"
}
```